utility/lambdas/ecs-asg-sizer/lambda_function.py

```diff
@@ -0,0 +1,115 @@
+import boto3
+import json
+import os
+
+session = boto3.session.Session()
+asgClient = session.client(service_name='autoscaling')
+ecsClient = session.client(service_name='ecs')
+
+#
+# Make an ASG the perfect size for an ECS cluster
+#
+def lambda_handler(event, context):
+    asgName = get_env('ASG_NAME')
+    clusterName = get_env('ECS_CLUSTER')
+    thresholdCpu = int(get_env('THRESHOLD_CPU'))
+    thresholdMemory = int(get_env('THRESHOLD_MEMORY'))
+    thresholdCount = int(get_env('THRESHOLD_COUNT'))
+
+    usages = get_usages(clusterName)
+    log_debug('Targeting threshold {0}x {1} cpu / {2} mem'.format(thresholdCount, thresholdCpu, thresholdMemory))
+    for usage in usages:
+        log_debug('    {0} - {1} / {2}'.format(usage['id'], usage['remainingCpu'], usage['remainingMemory']))
+    if len(usages) == 0:
+        log_error('No instances running in cluster {0}'.format())
+
+    # return early if already scaling
+    reason = is_scaling(asgName, len(usages))
+    if reason: log_debug('Already scaling: {0}'.format(reason)); return
+
+    # how many THRESHOLD-sized tasks can fit in this cluster
+    slotsAvailable = 0
+    for usage in usages:
+        cpu = usage['remainingCpu']
+        mem = usage['remainingMemory']
+        while cpu >= thresholdCpu and mem >= thresholdMemory:
+            slotsAvailable += 1
+            cpu -= thresholdCpu
+            mem -= thresholdMemory
+
+    # calculate how many THRESHOLD-sized tasks could fit on 1 instance
+    totalCpu = usages[0]['totalCpu']
+    totalMemory = usages[0]['totalMemory']
+    slotsPerInstance = 0
+    while totalCpu > 0 and totalMemory > 0:
+        slotsPerInstance += 1
+        totalCpu -= thresholdCpu
+        totalMemory -= thresholdMemory
+
+    # scale!
+    msg = '{0} slots available ({1}/instance)'.format(slotsAvailable, slotsPerInstance)
+    if slotsAvailable < thresholdCount:
+        log_info('Scale UP: ' + msg)
```

Comment:
We should standardize around up/down vs in/out. My preference would be in/out as that's what AWS seems to be using recently